hal/src/qnvm_ioblock.c

```diff
@@ -0,0 +1,328 @@
+    osalDbgAssert((nvmioblockp->state == BLK_STOP) || (nvmioblockp->state == BLK_READY),
+            "invalid state");
+
+    nvmioblockp->state = BLK_STOP;
+}
+
+/**
+ * @brief   Reads data crossing sector boundaries if required.
+ *
+ * @param[in] nvmioblockp   pointer to the @p NVMIOBlockDriver object
+ * @param[in] startblk      block to start reading from
+ * @param[in] buffer        pointer to data buffer
+ * @param[in] n             number of blocks to read
+ *
+ * @return                  The operation status.
+ * @retval HAL_SUCCESS      the operation succeeded.
+ * @retval HAL_FAILED       the operation failed.
+ *
+ * @api
+ */
+bool nvmioblockRead(NVMIOBlockDriver* nvmioblockp, uint32_t startblk,
+        uint8_t* buffer, uint32_t n)
+{
+    osalDbgCheck(nvmioblockp != NULL);
+    /* Verify device status. */
+    osalDbgAssert(nvmioblockp->state >= BLK_READY, "invalid state");
+
+    if (nvmioblockSync(nvmioblockp) != HAL_SUCCESS)
+        return HAL_FAILED;
+
+    /* Read operation in progress. */
+    nvmioblockp->state = BLK_READING;
+
+    bool result = nvmRead(nvmioblockp->config->nvmp,
+            nvmioblockp->config->block_size * startblk,
+            nvmioblockp->config->block_size * n,
+            buffer);
+
+    /* Read operation finished. */
+    nvmioblockp->state = BLK_READY;
+
+    return result;
+}
+
+/**
+ * @brief   Writes data crossing sector boundaries if required.
+ *
+ * @param[in] nvmioblockp   pointer to the @p NVMIOBlockDriver object
+ * @param[in] startblk      address to start writing to
+ * @param[in] buffer        pointer to data buffer
+ * @param[in] n             number of bytes to write
+ *
+ * @return                  The operation status.
+ * @retval HAL_SUCCESS      the operation succeeded.
+ * @retval HAL_FAILED       the operation failed.
+ *
+ * @api
+ */
+bool nvmioblockWrite(NVMIOBlockDriver* nvmioblockp, uint32_t startblk,
+        const uint8_t* buffer, uint32_t n)
+{
+    osalDbgCheck(nvmioblockp != NULL);
+    /* Verify device status. */
+    osalDbgAssert(nvmioblockp->state >= BLK_READY, "invalid state");
+
+    if (nvmioblockSync(nvmioblockp) != HAL_SUCCESS)
+        return HAL_FAILED;
+
+    /* Write operation in progress. */
+    nvmioblockp->state = BLK_WRITING;
+
+    bool result = nvmWrite(nvmioblockp->config->nvmp,
+            nvmioblockp->config->block_size * startblk,
+            nvmioblockp->config->block_size * n,
+            buffer);
+
+    /* Write operation finished. */
+    nvmioblockp->state = BLK_READY;
+
+    return result;
+}
+
+/**
+ * @brief   Waits for idle condition.
+ *
+ * @param[in] nvmioblockp   pointer to the @p NVMIOBlockDriver object
+ *
+ * @return                  The operation status.
+ * @retval HAL_SUCCESS      the operation succeeded.
+ * @retval HAL_FAILED       the operation failed.
+ *
+ * @api
+ */
+bool nvmioblockSync(NVMIOBlockDriver* nvmioblockp)
+{
+    osalDbgCheck(nvmioblockp != NULL);
+    /* Verify device status. */
```

Comment:
I do not like these  kind of comments. It is obvious to me what the following debug check does.